Scaladoc: under `-release`, fix generation of links to JVM classes

Math-ias opened the pull request:
Fixes scala/bug#12820.

som-snytt commented:
ct.sym has module info.
```
00000010  00 00 00 00 00 00 00 00  00 00 23 00 00 00 38 2f  |..........#...8/|
00000020  6a 61 76 61 2e 61 63 74  69 76 61 74 69 6f 6e 2f  |java.activation/|
00000030  6a 61 76 61 78 2f 61 63  74 69 76 61 74 69 6f 6e  |javax/activation|
00000040  2f 03 00 50 4b 07 08 00  00 00 00 02 00 00 00 00  |/..PK...........|
```
oh apparently I dumped it out already a year ago Halloween in my `/tmp/ct` which I learned only after trying that path again. So I've forgotten what I tried before.

I see `K/java.base` for example and `K/java.base/module-info.sig`.

This is where it indexes the available packages:
https://github.com/scala/scala/blob/v2.13.15/src/compiler/scala/tools/nsc/classpath/DirectoryClassPath.scala#L285

It could also preserve modules. The modules needn't be exposed through the ClassPath interface; Scaladoc could just use CtSym directly.

I remember this line to switch class path representations, so it's not unprecedented to be ct.sym-aware:
https://github.com/scala/scala/blob/v2.13.15/src/compiler/scala/tools/nsc/Global.scala#L150

I'm dubious about using `Class.forName` here, as that uses the tool's class loader. It might kind of work.

The tool itself must be JDK 8 compatible, per the CI failure.

Math-ias commented:
> ct.sym has module info.

I did not know the ct.sym file had module info, I will take a look at using it. I did find [the openjdk tool that makes that file for a reference](https://github.com/openjdk/jdk/blob/master/make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java).

Though parsing the `ct.sym` will crack only half the examples (better than none). I've come up with a different replication. https://github.com/scala/bug/issues/12820#issuecomment-2406023982

> I'm dubious about using Class.forName here, as that uses the tool's class loader. It might kind of work.

I tried to generalize my solution to also link external javadocs correctly. The external example class I tried couldn't be found on the tool's class loader but I didn't dig further. I didn't know the tool's class loader might have different access. I suppose it kind of worked for me in my limited tests because stdlib classes are generally present.

The java 8 requirement means I'll go back to the drawing board though. `ct.sym` or something else.

SethTisue commented:
> The tool itself must be JDK 8 compatible, per the CI failure

Sigh... that would be a sad constraint for us to have live under more or less forever (since it's not clear when or if Scala 2 might drop Java 8 support).

If we accept that constraint, then there's no way for us to do anything at all with JPMS.

Perhaps @lrytz would like to weigh in on this. If behaving reasonably on JDK 9+ requires us to write some JDK 9+ specific code, then I would hope we could find some way to accommodate that code in this project. I would suggest proceeding with your fix on the assumption that we will find a way to incorporate the fix if we need to, even if it doesn't compile on Java 8.

SethTisue commented:
> I did not know the ct.sym file had module info, I will take a look at using it

Hmm, so does that mean you're not currently stuck? I'm sorry I haven't offered help sooner (but I'm returning from a trifecta of illness, vacation, and sabbatical).

lrytz commented:
> If behaving reasonably on JDK 9+ requires us to write some JDK 9+ specific code

Right, so if the compiler runs on Java 8 there's no ct.sym file. We can access the necessary API reflectively if running on Java 9+, it seems to be only `Class.getModule()` and `module.getName()`.

SethTisue commented:
thanks, reflection is a good suggestion if the amount of code involved is small enough

Math-ias commented:
> no way for us to do anything at all with JPMS.
> so does that mean you're not currently stuck?

Yes in the meantime I've been researching how to read [java module information](https://docs.oracle.com/javase/specs/jvms/se23/html/jvms-4.html#jvms-4.4.11) from class files. This is relevant because `ct.sym` seems to be _roughly_ a JAR (zip archive) of _roughly_ class files (sig files).

With my current notes and the latest discussion I think should ...
 * Save reading java modules from class files for a different issue
 * Narrow fix with if statements to fix _just_ the original reported issue (would make me feel better about risk).
 * Use reflection to _try_ to access java module methods.
 * Create a new issue to fix this in java 8 (maybe where my module reading goes).

lrytz commented:
Is there anything to fix if scaladoc is running on Java 8? There are no modules there, so the API links are correct, no?

Math-ias commented:
@lrytz good point!
```
ChoiceSetting("-release", "release", "Compile for a version of the Java API and target class file.", AllTargetVersions, normalizeTarget(javaSpecVersion))
      .withPostSetHook { setting =>
        val current = setting.value.toInt
        if (!isJavaAtLeast("9") && current > 8) errorFn.apply("-release is only supported on JVM 9 and higher")
```
I'm just about ready. All that's left is linking to the targeted JDK's javadocs.

Math-ias commented:
I notice this "combined" step marked with an x saying a previous commit has failures. Does this mean I need to squash commits?

SethTisue commented:
Before we hit "merge" every commit needs to individually pass CI in sequence. You could rebase now, or wait, either way. (Sometimes it's easier for reviewers if you hold off on rebasing until the end, if there's been review feedback and then new commits that specifically address that feedback. Or you might decide you'd rather just have reviewers look at the cleaned-up version.)

SethTisue commented on src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:
```diff
@@ -60,9 +60,9 @@ trait MemberLookup extends base.MemberLookupBase {
         sym.info.member(newTermName("package"))
       else sym
-    classpathEntryFor(sym1) flatMap { path =>
+    classpathEntryFor(sym1) flatMap { case path =>
```
I don't see a reason to add this?

Math-ias replied:
Pushed to remove this. I was experimenting with if I could make the if statements clearer but decided against the change and forgot to remove all my changes.

SethTisue commented on src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:
```diff
@@ -100,11 +100,36 @@ trait MemberLookup extends base.MemberLookupBase {
 
-  def jdkUrl(path: String): String = {
-    if (path.endsWith(".jmod")) {
+  // ISSUE-12820
```
the new code beginning here is incorrectly indented, mind correcting that?

Math-ias replied:
Passed the new code into scalafmt like `pbpaste | scalafmt --stdin --stdout` to help me find and auto fix errors. Used the following `.scalafmt.conf`.
```
version = 3.8.4-RC1
runner.dialect = scala213
```

SethTisue replied:
I'm sorry, I was wrong

SethTisue commented on src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:
```diff
@@ -100,11 +100,35 @@ trait MemberLookup extends base.MemberLookupBase {
+  }.toOption
+
+  def jdkUrl(path: String, sym: Symbol): String = {
+    if (path.endsWith(".jmod") && javaVersion > 11) {
```
maybe I'm missing something obvious, but why `> 11` and not `>= 11` (or even `>= 9`)? Wasn't this stuff added in JDK 9?

I don't actually care about JDK 9 or 10, but I am wondering why 11 is excluded here.

Math-ias replied:
Oh my goodness, very embarrassing for me, it should be `>= 11`. I sneak it into one of commit messages, but javadocs _appear_ to start using the module name as part of the path starting in 11 even though modules are added in 9. I can't find a formal announcement of this change, just something I noticed while testing. Though it appears I didn't test enough. Compare links below ...

```
https://docs.oracle.com/en/java/javase/12/docs/api/java.base/java/lang/Integer.html
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Integer.html
https://docs.oracle.com/javase/10/docs/api/java/lang/Integer.html
https://docs.oracle.com/javase/9/docs/api/java/lang/Integer.html
https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html
```

Math-ias replied:
Changed the kind of equality used in the latest force push.

SethTisue replied:
(pats self on back and says "good eye" 😁 )

som-snytt commented on src/scaladoc/scala/tools/nsc/doc/model/MemberLookup.scala:
```diff
@@ -128,11 +152,15 @@ trait MemberLookup extends base.MemberLookupBase {
+      case Array(Some(1), Some(n)) => n // example: 1.8.0_242
+      case Array(Some(n))          => n // example: 14
+      case Array(Some(n), _)       => n // example: 11.0.7
+      case _                       => 8 // shrug!
```
Should use `scala.util.Properties` instead
```
scala> util.Properties.javaSpecVersion
val res0: String = 23

scala> util.Properties.isJavaAtLeast(11)
val res1: Boolean = true
```

Math-ias replied:
I've pushed changes that replace `System.getProperty("java.specification.version")` with `scala.util.Properties.javaSpecVersion`. I couldn't replace version checks in this file e.g ...

```
if (javaVersion < 11) {
      s"https://docs.oracle.com/javase/$javaVersion/docs/api"
```

with `isJavaAtLeast` calls. `javaSpecVersion` seems to be pinned to the version of the JVM I'm running the scaladoc tool with, whereas a lot of these checks are for what version of JVM we're targeting with the scaladoc tool (specified with `-release`).

Math-ias commented:
Thank you all for bearing with me and my mistakes in my first PR. 😄

Math-ias committed:
[ISSUE-12820] Try link to java module and the right javadoc ver.

When using release flag java loads stdlib classes from ct.sym file.
This means we can't properly figure out the java module of a class
from just the class path entry string itself.

I use reflection because I know java stdlib classes will be loadable.

Modules were added in java 9,
but didn't become part of the javadoc path until 11.
I couldn't find a formal announcement of this change.

Math-ias commented:
Force pushed to slightly clean up the author section, I had two username/email combos.

SethTisue commented:
Thank you!
